### PR TITLE
Reload window on memory out of bounds error

### PR DESCRIPTION
**Pull request by felipecrs**
https://user-images.githubusercontent.com/29582865/158080464-48ade8f0-1b1a-4fec-a8fe-9274dfb69107.mp4

Refs #388

### CODE REVIEW

**Comment by felipecrs** [comment]
Ops, just saw https://github.com/dermotduffy/frigate-hass-card/issues/388#issuecomment-1065906642.

@dermotduffy no worries if you don't accept this PR, I will understand. Meanwhile, I'll experiment with `FrigateCardLiveJSMPEG._refreshPlayer()` as well.

**Comment by dermotduffy** [comment]
Firstly definitely open to a change like this if we can make it:

* Solve this problem (even hackily) for those that have it. We should ask someone else to experiment also, since I don't seem to have this issue.
* Not need to reload the whole document (as per other comment).

Rather than using `patch-package`, did you try doing a bit of plumbing? If we could get it to work without an actual patch, that is preferred. Here is what I had been experimenting with:

* Make a subclass of `JSMpeg.Source['WebSocket']` ([ref](https://github.com/cycjimmy/jsmpeg-player/blob/master/src/index.js#L57))
* Do the catching / trickery in that subclass.
* Pass in the new "improved" source as an option, which is natively supported ([ref](https://github.com/cycjimmy/jsmpeg-player/blob/0f78d58000fe991397f60a6c5d6d97db776a294e/src/lib/player.js#L34)).

I suspect that's much less fragile than a literal patch, since these APIs are exposed by the player.

**Comment by felipecrs** [comment]
I'm not sure why, but:

```console
Uncaught TypeError: FrigateCardLiveJSMPEG._refreshPlayer is not a function
```

Although nothing seems wrong in the code.

**Comment by felipecrs** [comment]
Thanks for the suggestion, I'll try to do it later today. Thank you!

**Comment by dermotduffy** [comment]
> I'm not sure why, but:

How are you calling it? (Stating the likely obvious, it's not a static method so you need access to the `FrigateCardLiveJSMPEG` object to call it on, ideally it would be called from __within__ FrigateCardLiveJSMPEG somehow -- my 'plumbing' suggestion above may make that easier).

**Comment by felipecrs** [comment]
I called directly, since at some point in the code it's also being declared globally -- in the bundled code, of course:

```console
let FrigateCardLiveJSMPEG = [...]
```

But that's okay. You are right, if I manage to do it in the right way I can probably call it from within itself.

Thanks again!

**Comment by felipecrs** [comment]
I think I implemented what you had in mind, @dermotduffy, but something weird is going on now:

https://user-images.githubusercontent.com/29582865/158076080-365c8f4c-8a20-4250-811d-9cbe1416bed7.mp4

Not sure if there is something missing in the `_refreshPlayer()` routine, but I didn't find anything useful yet.

File: src/components/live.ts [hunk 1]
```diff
@@ -1103,7 +1106,7 @@ export class FrigateCardLiveJSMPEG extends LitElement {
   /**
    * Refresh the JSMPEG player.
    */
-  protected async _refreshPlayer(): Promise<void> {
+  async _refreshPlayer(): Promise<void> {
```
Comment (dermotduffy) [inline src/components/live.ts:1109]: Remove the leading `_` since you are making this method public.

File: src/components/live.ts [hunk 2]
```diff
@@ -1039,6 +1039,9 @@ export class FrigateCardLiveJSMPEG extends LitElement {
           // Override with user-specified options.
           ...this.jsmpegConfig?.options,
 
+          frigatePlayer: this,
```
Comment (dermotduffy) [inline src/components/live.ts:1042]: `frigateCardLivePlayer` ?

**Review by dermotduffy — COMMENTED** [review]
Yes, this is ~exactly what I was thinking of, thank you!

**Comment by dermotduffy** [comment]
> Not sure if there is something missing in the _refreshPlayer() routine, but I didn't find anything useful yet.

This smells of an infinite loop: the error is occurring, your new code is catching it, it refreshes the player, the error occurs, it refreshes the player, etc, etc. Each time it refreshes, it uses an additional GL context until that player is garbage collected. I suspect you're spawning players faster than Javascript is garbage collecting the old ones...

**Comment by felipecrs** [comment]
That would also mean, I guess, that refreshing the player isn't effective to prevent the underlying error from appearing.

File: src/components/live.ts (hunk 2 above)
Comment (felipecrs) [reply src/components/live.ts]: done.

File: src/components/live.ts (hunk 1 above)
Comment (felipecrs) [reply src/components/live.ts]: done.

**Comment by dermotduffy** [comment]
> That would also mean, I guess, that refreshing the player isn't effective to prevent the underlying error from appearing.

Yes, I think that's right :-(

**Comment by felipecrs** [comment]
I was able to fix the WebGL errors doing a kind of lock to prevent multiple `refreshPlayer()` calls from happening at the same time against the same player.

But with it, although I don't receive the WebGL errors anymore, it just tries to refresh infinitely anyway.

As we are ensuring that the error is a very specific one, we are sure to only reload window when strictly required.

I changed the code to do that instead.

If you want, I can use something like https://stackoverflow.com/a/41905026/12156188 to throw a warning in console after the window gets reloaded, so that users would at least get to know why their window got reloaded (which is much better than spinning forever anyway).

My only concern is about hooking on `window.onload`, I don't want to mess with something else in HA hooking on it.

**Comment by felipecrs** [comment]
PS: I added a demo in the PR description.

**Comment by felipecrs** [comment]
@maxi1134, @Mincka, @balonchiks, I took your names form #257 and #189. Is there any chance you can help me to test this approach?

Ensure you didn't disable WebAssembly (i.e. `jsmpeg.options.disableWebAssembly: true`), as it was one of the suggested workarounds of before.

Also, any feedback that you may have about the approach I used are welcome.

Thank you!

Here's how I do it:

https://user-images.githubusercontent.com/29582865/158091898-5fdcd02a-7a6e-4193-838b-54e1d3488a57.mp4

**Comment by felipecrs** [comment]
@dermotduffy I think I have great news. https://github.com/cycjimmy/jsmpeg-player/issues/27#issuecomment-1067662709 just updated the player (dependabot should send a PR soon).

And I haven't observed the issue ever since!

**Comment by dermotduffy** [comment]
Fantastic outcome @felipecrs . Thank you for chasing this upstream.

**Comment by felipecrs** [comment]
Thank you too for all the support, as always.

BTW: @dependabot won't send more than 5 updates as per the current configuration, so you should probably merge the existing ones first before it creates the one for the jsmpeg-player...

**Pull request closed without merging**